src/queue/multilevel.rs

```diff
@@ -0,0 +1,838 @@
+    fn prepare_before_push<T>(&self, task_cell: &mut T)
+    where
+        T: TaskCell,
+    {
+        let extras = task_cell.mut_extras();
+        let task_id = extras.task_id;
+        let running_time = extras
+            .running_time
+            .get_or_insert_with(|| self.task_elapsed_map.get_elapsed(task_id));
+        let current_level = match extras.fixed_level {
+            Some(level) => level,
+            None => {
+                let running_time = running_time.as_duration();
+                self.level_time_threshold
+                    .iter()
+                    .enumerate()
+                    .find(|(_, &threshold)| running_time < threshold)
+                    .map(|(level, _)| level)
+                    .unwrap_or(LEVEL_NUM - 1) as u8
+            }
+        };
+        extras.current_level = current_level;
+        extras.schedule_time = Some(now());
+    }
+
+    fn get_level0_chance(&self) -> u32 {
+        self.level0_chance.load(SeqCst)
+    }
+
+    fn maybe_adjust_chance(&self) {
+        if self.adjusting.compare_and_swap(false, true, SeqCst) {
+            return;
+        }
+        // The statistics may be not so accurate because we cannot load two
+        // atomics at the same time.
+        let total = self.total_elapsed.0.load(SeqCst);
+        if Duration::from_micros(total) < ADJUST_CHANCE_INTERVAL {
+            // Another thread just adjusted the chances.
+            return;
+        }
+        let total = self.total_elapsed.0.swap(0, SeqCst);
+        let level0 = self.level0_elapsed.0.swap(0, SeqCst);
+        let current_proportion = level0 as f64 / total as f64;
+        let diff = self.level0_proportion_target - current_proportion;
+        let level0_chance = self.level0_chance.load(SeqCst);
+        let new_chance = if diff > ADJUST_CHANCE_THRESHOLD {
+            cmp::min(
+                level0_chance.saturating_add(ADJUST_AMOUNT),
+                MAX_LEVEL0_CHANCE,
+            )
+        } else if diff < -ADJUST_CHANCE_THRESHOLD {
+            cmp::max(level0_chance - ADJUST_AMOUNT, MIN_LEVEL0_CHANCE)
+        } else {
+            level0_chance
+        };
+        self.level0_chance.store(new_chance, SeqCst);
+        self.adjusting.store(false, SeqCst);
+    }
+}
+
+#[derive(Default, Debug)]
+pub(crate) struct ElapsedTime(AtomicU64);
+
+impl ElapsedTime {
+    fn as_duration(&self) -> Duration {
+        Duration::from_micros(self.0.load(SeqCst))
+    }
+
+    fn inc_by(&self, t: Duration) -> u64 {
+        self.0.fetch_add(t.as_micros() as u64, SeqCst)
+    }
+
+    #[cfg(test)]
+    fn from_duration(dur: Duration) -> Self {
+        ElapsedTime(AtomicU64::new(dur.as_micros() as u64))
+    }
+}
+
+thread_local!(static TLS_LAST_CLEANUP_TIME: Cell<Instant> = Cell::new(Instant::now()));
+
+struct TaskElapsedMap {
+    new_index: AtomicUsize,
+    maps: [DashMap<u64, Arc<ElapsedTime>>; 2],
+    cleanup_interval: Duration,
+    last_cleanup_time: RwLock<Instant>,
```

Comment:
`Mutex` is better.

Reply:
Why is `Mutex` better? Multiple threads may read the value but only one will modify it. Using a `RwLock` enables multiple threads to read the value at the same time.

Reply:
Because `Mutex` is faster than `RwLock`, unless read lock is held for a long time, `Mutex` should always have the best performance.